server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java

```diff
@@ -19,204 +19,131 @@
-                .then(Mono.just(blobId)));
-    }
 
-    private Mono<Integer> saveBlobParts(BucketName bucketName, byte[] data, BlobId blobId) {
-        Stream<Pair<Integer, ByteBuffer>> chunks = dataChunker.chunk(data, configuration.getBlobPartSize());
-        return Flux.fromStream(chunks)
-            .publishOn(Schedulers.elastic(), PREFETCH)
-            .flatMap(pair -> writePart(bucketName, blobId, pair.getKey(), pair.getValue())
-                .then(Mono.just(getChunkNum(pair))))
-            .collect(Collectors.maxBy(Comparator.comparingInt(x -> x)))
-            .flatMap(Mono::justOrEmpty)
-            .map(this::numToCount)
-            .defaultIfEmpty(0);
+        return dumbBlobStore.save(bucketName, blobId, data)
+            .then(Mono.just(blobId));
     }
 
-
-    private int numToCount(int number) {
-        return number + 1;
-    }
-
-    private Integer getChunkNum(Pair<Integer, ByteBuffer> pair) {
-        return pair.getKey();
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(bucketName);
+        Preconditions.checkNotNull(data);
+        HashingInputStream hashingInputStream = new HashingInputStream(Hashing.sha256(), data);
+        return Mono.using(
+            () -> new FileBackedOutputStream(FILE_THRESHOLD),
+            fileBackedOutputStream -> saveAndGenerateBlobId(bucketName, hashingInputStream, fileBackedOutputStream),
+            fileBackedOutputStream -> {
+                try {
+                    fileBackedOutputStream.reset();
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            },
+            LAZY_RESSOURCE_CLEANUP);
+    }
+
+    private Mono<BlobId> saveAndGenerateBlobId(BucketName bucketName, HashingInputStream hashingInputStream, FileBackedOutputStream fileBackedOutputStream) {
+        return Mono.fromCallable(() -> {
+            IOUtils.copy(hashingInputStream, fileBackedOutputStream);
+            return Tuples.of(blobIdFactory.from(hashingInputStream.hash().toString()), fileBackedOutputStream.asByteSource());
+        })
+            .flatMap(tuple -> dumbBlobStore.save(bucketName, tuple.getT1(), tuple.getT2()).thenReturn(tuple.getT1()));
     }
 
     @Override
     public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
-        return readBlobParts(bucketName, blobId)
-            .collectList()
-            .map(this::byteBuffersToBytesArray);
+        Preconditions.checkNotNull(bucketName);
+        return retrieveStoredValue(bucketName, blobId);
     }
 
     @Override
     public InputStream read(BucketName bucketName, BlobId blobId) {
-        return ReactorUtils.toInputStream(readBlobParts(bucketName, blobId));
+        Preconditions.checkNotNull(bucketName);
+        return retrieveStoredValue(bucketName, blobId)
+            .map(ByteArrayInputStream::new)
```

Comment:
Are we loosing the `streaming` read capability in the equation?

Reply:
yes, this method is weird, needs to be fixed